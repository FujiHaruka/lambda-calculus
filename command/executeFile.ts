import { color } from "../deps.ts";
import { Repl } from "./repl.ts";

export function executeFile(
  script: string,
  print: (str: string) => void = console.log,
): number {
  const repl = new Repl();

  const lines = script.split("\n")
    .map((line) => line.trim())
    .filter(Boolean);

  for (const line of lines) {
    print(`> ${line}`);
    const out = repl.eval(line);
    if (out) {
      print(out);
    }
    if (out.startsWith(color.red("Error"))) {
      return 1;
    }
  }
  return 0;
}
