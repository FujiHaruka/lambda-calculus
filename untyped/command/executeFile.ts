import { Repl } from "./repl.ts";

export function executeFile(script: string): void {
  const repl = new Repl();

  const lines = script.split("\n")
    .map((line) => line.trim())
    .filter(Boolean);

  for (const line of lines) {
    console.log(`> ${line}`);
    const out = repl.eval(line);
    if (out) {
      console.log(out);
    }
  }
}
