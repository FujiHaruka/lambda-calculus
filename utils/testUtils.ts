export {
  assertEquals,
  assertStrictEquals,
  assertStringIncludes,
  assertThrows,
} from "https://deno.land/std@0.192.0/testing/asserts.ts";
export { describe, it } from "https://deno.land/std@0.192.0/testing/bdd.ts";
export { assertSnapshot } from "https://deno.land/std@0.192.0/testing/snapshot.ts";
export { walkSync } from "https://deno.land/std@0.192.0/fs/mod.ts";
export * as path from "https://deno.land/std@0.192.0/path/mod.ts";
