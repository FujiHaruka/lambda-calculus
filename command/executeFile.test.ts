import { assertSnapshot, it, path, walkSync } from "../testUtils.ts";
import { executeFile } from "./executeFile.ts";

const fixtureDir = path.resolve(
  path.dirname(path.fromFileUrl(import.meta.url)),
  "../fixtures",
);
const fixtures = walkSync(fixtureDir, { maxDepth: 1, exts: [".lambda"] });

for (const fixture of fixtures) {
  const fileName = path.basename(fixture.path);
  it(`executes ${fileName}`, async (t) => {
    const code = await Deno.readTextFile(fixture.path);
    const consoleOutput: string[] = [];
    const print = (str: string) => consoleOutput.push(str);
    executeFile(code, print);
    await assertSnapshot(t, consoleOutput);
  });
}
