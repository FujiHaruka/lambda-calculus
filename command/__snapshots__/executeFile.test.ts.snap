export const snapshot = {};

snapshot[`executes error.lambda 1`] = `
[
  "> # Error: Unexpected token",
  "> a -> -> b",
  "\\x1b[31mError\\x1b[39m: Unexpected token at position 5:

  a -> -> b
       ~~
",
]
`;

snapshot[`executes number.lambda 1`] = `
[
  "> \$x = REDUCE \$IS_ZERO \$0",
  "((n -> ((n (k -> (t -> (f -> f)))) (t -> (f -> t)))) (s -> (z -> z)))
" +
    "(((s -> (z -> z)) (k -> (t -> (f -> f)))) (t -> (f -> t)))
" +
    "((z -> z) (t -> (f -> t)))
" +
    "(t -> (f -> t))",
  "> ALPHA_EQ \$x, \$TRUE",
  "Yes",
  "> \$x = REDUCE \$IS_ZERO \$1",
  "((n -> ((n (k -> (t -> (f -> f)))) (t -> (f -> t)))) (s -> (z -> (s z))))
" +
    "(((s -> (z -> (s z))) (k -> (t -> (f -> f)))) (t -> (f -> t)))
" +
    "((z -> ((k -> (t -> (f -> f))) z)) (t -> (f -> t)))
" +
    "((k -> (t -> (f -> f))) (t -> (f -> t)))
" +
    "(t -> (f -> f))",
  "> ALPHA_EQ \$x, \$FALSE",
  "Yes",
  "> \$x = REDUCE \$SUCC \$0",
  "((n -> (s -> (z -> (s ((n s) z))))) (s -> (z -> z)))
" +
    "(s -> (z -> (s (((s -> (z -> z)) s) z))))
" +
    "(s -> (z -> (s ((z -> z) z))))
" +
    "(s -> (z -> (s z)))",
  "> ALPHA_EQ \$x, \$1",
  "Yes",
  "> \$x = REDUCE \$SUCC \$1",
  "((n -> (s -> (z -> (s ((n s) z))))) (s -> (z -> (s z))))
" +
    "(s -> (z -> (s (((s -> (z -> (s z))) s) z))))
" +
    "(s -> (z -> (s ((z -> (s z)) z))))
" +
    "(s -> (z -> (s (s z))))",
  "> ALPHA_EQ \$x, \$2",
  "Yes",
  "> \$x = REDUCE \$PLUS \$1 \$1",
  "(((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s z)))) (s -> (z -> (s z))))
" +
    "((n -> (s -> (z -> (((s -> (z -> (s z))) s) ((n s) z))))) (s -> (z -> (s z))))
" +
    "(s -> (z -> (((s -> (z -> (s z))) s) (((s -> (z -> (s z))) s) z))))
" +
    "(s -> (z -> ((z -> (s z)) (((s -> (z -> (s z))) s) z))))
" +
    "(s -> (z -> (s (((s -> (z -> (s z))) s) z))))
" +
    "(s -> (z -> (s ((z -> (s z)) z))))
" +
    "(s -> (z -> (s (s z))))",
  "> ALPHA_EQ \$x, \$2",
  "Yes",
  "> \$x = REDUCE \$PLUS \$2 \$3",
  "(((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> (s (s (s z))))))
" +
    "((n -> (s -> (z -> (((s -> (z -> (s (s z)))) s) ((n s) z))))) (s -> (z -> (s (s (s z))))))
" +
    "(s -> (z -> (((s -> (z -> (s (s z)))) s) (((s -> (z -> (s (s (s z))))) s) z))))
" +
    "(s -> (z -> ((z -> (s (s z))) (((s -> (z -> (s (s (s z))))) s) z))))
" +
    "(s -> (z -> (s (s (((s -> (z -> (s (s (s z))))) s) z)))))
" +
    "(s -> (z -> (s (s ((z -> (s (s (s z)))) z)))))
" +
    "(s -> (z -> (s (s (s (s (s z)))))))",
  "> ALPHA_EQ \$x, \$5",
  "Yes",
  "> \$x = REDUCE \$PLUS \$2 \$3",
  "(((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> (s (s (s z))))))
" +
    "((n -> (s -> (z -> (((s -> (z -> (s (s z)))) s) ((n s) z))))) (s -> (z -> (s (s (s z))))))
" +
    "(s -> (z -> (((s -> (z -> (s (s z)))) s) (((s -> (z -> (s (s (s z))))) s) z))))
" +
    "(s -> (z -> ((z -> (s (s z))) (((s -> (z -> (s (s (s z))))) s) z))))
" +
    "(s -> (z -> (s (s (((s -> (z -> (s (s (s z))))) s) z)))))
" +
    "(s -> (z -> (s (s ((z -> (s (s (s z)))) z)))))
" +
    "(s -> (z -> (s (s (s (s (s z)))))))",
  "> ALPHA_EQ \$x, \$5",
  "Yes",
  "> \$x = REDUCE \$MULT \$2 \$3",
  "(((m -> (n -> ((n ((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) m)) (s -> (z -> z))))) (s -> (z -> (s (s z))))) (s -> (z -> (s (s (s z))))))
" +
    "((n -> ((n ((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z)))))) (s -> (z -> z)))) (s -> (z -> (s (s (s z))))))
" +
    "(((s -> (z -> (s (s (s z))))) ((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z)))))) (s -> (z -> z)))
" +
    "((z -> (((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) z)))) (s -> (z -> z)))
" +
    "(((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> z)))))
" +
    "((n -> (s -> (z -> (((s -> (z -> (s (s z)))) s) ((n s) z))))) (((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> z)))))
" +
    "(s -> (z -> (((s -> (z -> (s (s z)))) s) (((((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> z)))) s) z))))
" +
    "(s -> (z -> ((z -> (s (s z))) (((((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> z)))) s) z))))
" +
    "(s -> (z -> (s (s (((((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> z)))) s) z)))))
" +
    "(s -> (z -> (s (s ((((n -> (s -> (z -> (((s -> (z -> (s (s z)))) s) ((n s) z))))) (((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> z)))) s) z)))))
" +
    "(s -> (z -> (s (s (((s -> (z -> (((s -> (z -> (s (s z)))) s) (((((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> z))) s) z)))) s) z)))))
" +
    "(s -> (z -> (s (s ((z -> (((s -> (z -> (s (s z)))) s) (((((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> z))) s) z))) z)))))
" +
    "(s -> (z -> (s (s (((s -> (z -> (s (s z)))) s) (((((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> z))) s) z))))))
" +
    "(s -> (z -> (s (s ((z -> (s (s z))) (((((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> z))) s) z))))))
" +
    "(s -> (z -> (s (s (s (s (((((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s (s z))))) (s -> (z -> z))) s) z)))))))
" +
    "(s -> (z -> (s (s (s (s ((((n -> (s -> (z -> (((s -> (z -> (s (s z)))) s) ((n s) z))))) (s -> (z -> z))) s) z)))))))
" +
    "(s -> (z -> (s (s (s (s (((s -> (z -> (((s -> (z -> (s (s z)))) s) (((s -> (z -> z)) s) z)))) s) z)))))))
" +
    "(s -> (z -> (s (s (s (s ((z -> (((s -> (z -> (s (s z)))) s) (((s -> (z -> z)) s) z))) z)))))))
" +
    "(s -> (z -> (s (s (s (s (((s -> (z -> (s (s z)))) s) (((s -> (z -> z)) s) z))))))))
" +
    "(s -> (z -> (s (s (s (s ((z -> (s (s z))) (((s -> (z -> z)) s) z))))))))
" +
    "(s -> (z -> (s (s (s (s (s (s (((s -> (z -> z)) s) z)))))))))
" +
    "(s -> (z -> (s (s (s (s (s (s ((z -> z) z)))))))))
" +
    "(s -> (z -> (s (s (s (s (s (s z))))))))",
  "> ALPHA_EQ \$x, \$6",
  "Yes",
  "> \$x = REDUCE \$PRED \$1",
  "((n -> (f -> (x -> (((n (g -> (h -> (h (g f))))) (u -> x)) (u -> u))))) (s -> (z -> (s z))))
" +
    "(f -> (x -> ((((s -> (z -> (s z))) (g -> (h -> (h (g f))))) (u -> x)) (u -> u))))
" +
    "(f -> (x -> (((z -> ((g -> (h -> (h (g f)))) z)) (u -> x)) (u -> u))))
" +
    "(f -> (x -> (((g -> (h -> (h (g f)))) (u -> x)) (u -> u))))
" +
    "(f -> (x -> ((h -> (h ((u -> x) f))) (u -> u))))
" +
    "(f -> (x -> ((u -> u) ((u -> x) f))))
" +
    "(f -> (x -> ((u -> x) f)))
" +
    "(f -> (x -> x))",
  "> ALPHA_EQ \$x, \$0",
  "Yes",
  "> \$x = REDUCE \$SUB \$5 \$2",
  "(((m -> (n -> ((n (n -> (f -> (x -> (((n (g -> (h -> (h (g f))))) (u -> x)) (u -> u)))))) m))) (s -> (z -> (s (s (s (s (s z)))))))) (s -> (z -> (s (s z)))))
" +
    "((n -> ((n (n -> (f -> (x -> (((n (g -> (h -> (h (g f))))) (u -> x)) (u -> u)))))) (s -> (z -> (s (s (s (s (s z))))))))) (s -> (z -> (s (s z)))))
" +
    "(((s -> (z -> (s (s z)))) (n -> (f -> (x -> (((n (g -> (h -> (h (g f))))) (u -> x)) (u -> u)))))) (s -> (z -> (s (s (s (s (s z))))))))
" +
    "((z -> ((n -> (f -> (x -> (((n (g -> (h -> (h (g f))))) (u -> x)) (u -> u))))) ((n -> (f -> (x -> (((n (g -> (h -> (h (g f))))) (u -> x)) (u -> u))))) z))) (s -> (z -> (s (s (s (s (s z))))))))
" +
    "((n -> (f -> (x -> (((n (g -> (h -> (h (g f))))) (u -> x)) (u -> u))))) ((n -> (f -> (x -> (((n (g -> (h -> (h (g f))))) (u -> x)) (u -> u))))) (s -> (z -> (s (s (s (s (s z)))))))))
" +
    "(f -> (x -> (((((n -> (f -> (x -> (((n (g -> (h -> (h (g f))))) (u -> x)) (u -> u))))) (s -> (z -> (s (s (s (s (s z)))))))) (g -> (h -> (h (g f))))) (u -> x)) (u -> u))))
" +
    "(f -> (x -> ((((f -> (x -> ((((s -> (z -> (s (s (s (s (s z))))))) (g -> (h -> (h (g f))))) (u -> x)) (u -> u)))) (g -> (h -> (h (g f))))) (u -> x)) (u -> u))))
" +
    "(f -> (x -> (((x -> ((((s -> (z -> (s (s (s (s (s z))))))) (g -> (h -> (h (g (g -> (h -> (h (g f))))))))) (u -> x)) (u -> u))) (u -> x)) (u -> u))))
" +
    "(f -> (x -> (((((s -> (z -> (s (s (s (s (s z))))))) (g -> (h -> (h (g (g -> (h -> (h (g f))))))))) (u -> (u -> x))) (u -> u)) (u -> u))))
" +
    "(f -> (x -> ((((z -> ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) z)))))) (u -> (u -> x))) (u -> u)) (u -> u))))
" +
    "(f -> (x -> ((((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x))))))) (u -> u)) (u -> u))))
" +
    "(f -> (x -> (((h -> (h (((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x)))))) (g -> (h -> (h (g f))))))) (u -> u)) (u -> u))))
" +
    "(f -> (x -> (((u -> u) (((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x)))))) (g -> (h -> (h (g f)))))) (u -> u))))
" +
    "(f -> (x -> ((((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x)))))) (g -> (h -> (h (g f))))) (u -> u))))
" +
    "(f -> (x -> (((h -> (h (((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x))))) (g -> (h -> (h (g f))))))) (g -> (h -> (h (g f))))) (u -> u))))
" +
    "(f -> (x -> (((g -> (h -> (h (g f)))) (((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x))))) (g -> (h -> (h (g f)))))) (u -> u))))
" +
    "(f -> (x -> ((h -> (h ((((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x))))) (g -> (h -> (h (g f))))) f))) (u -> u))))
" +
    "(f -> (x -> ((u -> u) ((((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x))))) (g -> (h -> (h (g f))))) f))))
" +
    "(f -> (x -> ((((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x))))) (g -> (h -> (h (g f))))) f)))
" +
    "(f -> (x -> (((h -> (h (((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x)))) (g -> (h -> (h (g f))))))) (g -> (h -> (h (g f))))) f)))
" +
    "(f -> (x -> (((g -> (h -> (h (g f)))) (((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x)))) (g -> (h -> (h (g f)))))) f)))
" +
    "(f -> (x -> ((h -> (h ((((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x)))) (g -> (h -> (h (g f))))) f))) f)))
" +
    "(f -> (x -> (f ((((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) ((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x)))) (g -> (h -> (h (g f))))) f))))
" +
    "(f -> (x -> (f (((h -> (h (((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x))) (g -> (h -> (h (g f))))))) (g -> (h -> (h (g f))))) f))))
" +
    "(f -> (x -> (f (((g -> (h -> (h (g f)))) (((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x))) (g -> (h -> (h (g f)))))) f))))
" +
    "(f -> (x -> (f ((h -> (h ((((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x))) (g -> (h -> (h (g f))))) f))) f))))
" +
    "(f -> (x -> (f (f ((((g -> (h -> (h (g (g -> (h -> (h (g f)))))))) (u -> (u -> x))) (g -> (h -> (h (g f))))) f)))))
" +
    "(f -> (x -> (f (f (((h -> (h ((u -> (u -> x)) (g -> (h -> (h (g f))))))) (g -> (h -> (h (g f))))) f)))))
" +
    "(f -> (x -> (f (f (((g -> (h -> (h (g f)))) ((u -> (u -> x)) (g -> (h -> (h (g f)))))) f)))))
" +
    "(f -> (x -> (f (f ((h -> (h (((u -> (u -> x)) (g -> (h -> (h (g f))))) f))) f)))))
" +
    "(f -> (x -> (f (f (f (((u -> (u -> x)) (g -> (h -> (h (g f))))) f))))))
" +
    "(f -> (x -> (f (f (f ((u -> x) f))))))
" +
    "(f -> (x -> (f (f (f x)))))",
  "> ALPHA_EQ \$x, \$3",
  "Yes",
]
`;

snapshot[`executes simple.lambda 1`] = `
[
  "> \$id = x -> x",
  "(x -> x)",
  '> # Reduced to "f"',
  "> REDUCE \$id f",
  "((x -> x) f)
f",
  "> \$y  = x -> y",
  "(x -> y)",
  '> # Reduced to "y"',
  "> REDUCE \$y f",
  "((x -> y) f)
y",
]
`;

snapshot[`executes alpha_eq.lambda 1`] = `
[
  "> # Yes",
  "> ALPHA_EQ x -> x, t -> t",
  "Yes",
  "> # No",
  "> ALPHA_EQ x -> y, x -> t",
  "No",
  "> \$n = REDUCE \$PLUS \$1 \$1",
  "(((m -> (n -> (s -> (z -> ((m s) ((n s) z)))))) (s -> (z -> (s z)))) (s -> (z -> (s z))))
" +
    "((n -> (s -> (z -> (((s -> (z -> (s z))) s) ((n s) z))))) (s -> (z -> (s z))))
" +
    "(s -> (z -> (((s -> (z -> (s z))) s) (((s -> (z -> (s z))) s) z))))
" +
    "(s -> (z -> ((z -> (s z)) (((s -> (z -> (s z))) s) z))))
" +
    "(s -> (z -> (s (((s -> (z -> (s z))) s) z))))
" +
    "(s -> (z -> (s ((z -> (s z)) z))))
" +
    "(s -> (z -> (s (s z))))",
  "> # Yes",
  "> ALPHA_EQ \$n, \$2",
  "Yes",
]
`;

snapshot[`executes bool.lambda 1`] = `
[
  "> \$TRUE",
  "(t -> (f -> t))",
  "> \$FALSE",
  "(t -> (f -> f))",
  "> REDUCE \$NOT \$TRUE",
  "((p -> ((p (t -> (f -> f))) (t -> (f -> t)))) (t -> (f -> t)))
" +
    "(((t -> (f -> t)) (t -> (f -> f))) (t -> (f -> t)))
" +
    "((f -> (t -> (f -> f))) (t -> (f -> t)))
" +
    "(t -> (f -> f))",
  "> REDUCE \$NOT \$FALSE",
  "((p -> ((p (t -> (f -> f))) (t -> (f -> t)))) (t -> (f -> f)))
" +
    "(((t -> (f -> f)) (t -> (f -> f))) (t -> (f -> t)))
" +
    "((f -> f) (t -> (f -> t)))
" +
    "(t -> (f -> t))",
  "> REDUCE \$AND \$TRUE \$TRUE",
  "(((p -> (q -> ((p q) (t -> (f -> f))))) (t -> (f -> t))) (t -> (f -> t)))
" +
    "((q -> (((t -> (f -> t)) q) (t -> (f -> f)))) (t -> (f -> t)))
" +
    "(((t -> (f -> t)) (t -> (f -> t))) (t -> (f -> f)))
" +
    "((f -> (t -> (f -> t))) (t -> (f -> f)))
" +
    "(t -> (f -> t))",
  "> REDUCE \$AND \$TRUE \$FALSE",
  "(((p -> (q -> ((p q) (t -> (f -> f))))) (t -> (f -> t))) (t -> (f -> f)))
" +
    "((q -> (((t -> (f -> t)) q) (t -> (f -> f)))) (t -> (f -> f)))
" +
    "(((t -> (f -> t)) (t -> (f -> f))) (t -> (f -> f)))
" +
    "((f -> (t -> (f -> f))) (t -> (f -> f)))
" +
    "(t -> (f -> f))",
  "> REDUCE \$AND \$FALSE \$TRUE",
  "(((p -> (q -> ((p q) (t -> (f -> f))))) (t -> (f -> f))) (t -> (f -> t)))
" +
    "((q -> (((t -> (f -> f)) q) (t -> (f -> f)))) (t -> (f -> t)))
" +
    "(((t -> (f -> f)) (t -> (f -> t))) (t -> (f -> f)))
" +
    "((f -> f) (t -> (f -> f)))
" +
    "(t -> (f -> f))",
  "> REDUCE \$AND \$FALSE \$FALSE",
  "(((p -> (q -> ((p q) (t -> (f -> f))))) (t -> (f -> f))) (t -> (f -> f)))
" +
    "((q -> (((t -> (f -> f)) q) (t -> (f -> f)))) (t -> (f -> f)))
" +
    "(((t -> (f -> f)) (t -> (f -> f))) (t -> (f -> f)))
" +
    "((f -> f) (t -> (f -> f)))
" +
    "(t -> (f -> f))",
  "> REDUCE \$OR \$TRUE \$TRUE",
  "(((p -> (q -> ((p (t -> (f -> t))) q))) (t -> (f -> t))) (t -> (f -> t)))
" +
    "((q -> (((t -> (f -> t)) (t -> (f -> t))) q)) (t -> (f -> t)))
" +
    "(((t -> (f -> t)) (t -> (f -> t))) (t -> (f -> t)))
" +
    "((f -> (t -> (f -> t))) (t -> (f -> t)))
" +
    "(t -> (f -> t))",
  "> REDUCE \$OR \$TRUE \$FALSE",
  "(((p -> (q -> ((p (t -> (f -> t))) q))) (t -> (f -> t))) (t -> (f -> f)))
" +
    "((q -> (((t -> (f -> t)) (t -> (f -> t))) q)) (t -> (f -> f)))
" +
    "(((t -> (f -> t)) (t -> (f -> t))) (t -> (f -> f)))
" +
    "((f -> (t -> (f -> t))) (t -> (f -> f)))
" +
    "(t -> (f -> t))",
  "> REDUCE \$OR \$FALSE \$TRUE",
  "(((p -> (q -> ((p (t -> (f -> t))) q))) (t -> (f -> f))) (t -> (f -> t)))
" +
    "((q -> (((t -> (f -> f)) (t -> (f -> t))) q)) (t -> (f -> t)))
" +
    "(((t -> (f -> f)) (t -> (f -> t))) (t -> (f -> t)))
" +
    "((f -> f) (t -> (f -> t)))
" +
    "(t -> (f -> t))",
  "> REDUCE \$OR \$FALSE \$FALSE",
  "(((p -> (q -> ((p (t -> (f -> t))) q))) (t -> (f -> f))) (t -> (f -> f)))
" +
    "((q -> (((t -> (f -> f)) (t -> (f -> t))) q)) (t -> (f -> f)))
" +
    "(((t -> (f -> f)) (t -> (f -> t))) (t -> (f -> f)))
" +
    "((f -> f) (t -> (f -> f)))
" +
    "(t -> (f -> f))",
]
`;
